server/re/src/msi_genquery2.cpp

```diff
@@ -0,0 +1,213 @@
+            ++ctx.current_row;
+            log_msi::trace("{}: Incremented row position [{} => {}]. Returning 0.", __func__, ctx.current_row - 1, ctx.current_row);
+            return 0;
+        }
+
+        log_msi::trace("{}: Skipping increment of row position [current_row=[{}]]. Returning 1.", __func__, ctx.current_row);
+
+        // TODO Update this.
+        // We must return ERROR(stop_code, "") to trigger correct usage of msi_genquery2_next_row().
+        // Otherwise, the NREP can loop forever. Ultimately, this means we aren't allowed to return
+        // CODE(stop_code) to signal there is no new row available.
+        return 1;
+    }
+    catch (const irods::exception& e) {
+        log_msi::error("{}: {}", __func__, e.client_display_what());
+        return static_cast<int>(e.code());
+    }
+    catch (const std::exception& e) {
+        log_msi::error("{}: {}", __func__, e.what());
+        return SYS_LIBRARY_ERROR;
+    }
+
+    return 0;
+} // msi_genquery2_next_row
+
+auto msi_genquery2_column(MsParam* _handle, MsParam* _column_index, MsParam* _column_value, RuleExecInfo* _rei) -> int
+{
+    log_msi::trace(__func__);
+
+    IRODS_MSI_REQUIRE_VALID_POINTER(_handle);
+    IRODS_MSI_REQUIRE_VALID_POINTER(_column_index);
+    IRODS_MSI_REQUIRE_VALID_POINTER(_column_value);
+    IRODS_MSI_REQUIRE_VALID_POINTER(_rei);
+
+    IRODS_MSI_REQUIRE_VALID_POINTER(_handle->type);
+    IRODS_MSI_REQUIRE_VALID_POINTER(_column_index->type);
+
+    IRODS_MSI_REQUIRE_TYPE(_handle->type, STR_MS_T);
+    IRODS_MSI_REQUIRE_TYPE(_column_index->type, STR_MS_T);
+
+    IRODS_MSI_REQUIRE_VALID_POINTER(_handle->inOutStruct);
+    IRODS_MSI_REQUIRE_VALID_POINTER(_column_index->inOutStruct);
+
+    try {
+        const auto ctx_handle_index = std::stoll(static_cast<char*>(_handle->inOutStruct));
+
+        if (ctx_handle_index < 0 ||
+                static_cast<decltype(gq2_context)::size_type>(ctx_handle_index) >= gq2_context.size()) {
+            log_msi::error("{}: Unknown context handle.", __func__);
+            return SYS_INVALID_INPUT_PARAM;
+        }
+
+        auto& ctx = gq2_context.at(static_cast<decltype(gq2_context)::size_type>(ctx_handle_index));
+        const auto column_index = std::stoll(static_cast<char*>(_column_index->inOutStruct));
+
+        const auto& value = ctx.rows.at(ctx.current_row).at(column_index).get_ref<const std::string&>();
+        log_msi::debug("{}: Column value = [{}]", __func__, value);
+        fillStrInMsParam(_column_value, value.c_str());
+
+        return 0;
+    }
+    catch (const irods::exception& e) {
+        log_msi::error("{}: {}", __func__, e.client_display_what());
+        return static_cast<int>(e.code());
+    }
+    catch (const std::exception& e) {
+        log_msi::error("{}: {}", __func__, e.what());
+        return SYS_LIBRARY_ERROR;
+    }
+
+    return 0;
+} // msi_genquery2_column
+
+auto msi_genquery2_free(MsParam* _handle, RuleExecInfo* _rei) -> int
```

Comment:
Heads up!

This was called `msi_genquery2_destroy` in the original implementation.

Reply:
This is fine.